### PR TITLE
[WIP] Symfony authentication

### PR DESCRIPTION
**Pull request by bytehead**
This is an early working draft to use the replace the current `simple_preauth` mechanism with the Symfony `form_login` authentication listener.

Please review and mention all missed topics.

ToDos:
- [ ] Fix messages on `be_login.html5`
- [ ] Implement `$GLOBALS['TL_HOOKS']['postAuthenticate']` as an event
- [ ] Test frontend preview
- [ ] ~~Use `form_login` for `FrontendUser` (members) as well~~*
- [ ] ~~Implement role switcher / impersonate any `FrontendUser` for frontend preview~~*
- [ ] Write tests
- [ ] Code style

*I'll provide the Symfony authentication for FE users in a separate PR.

### CODE REVIEW

**Comment by aschempp** [comment]
❤️ 

I think we should not add frontend yet, just replace the backend in 4.3. This will give use more time to see into the solution, as frontend is more complex to add and has more potential side effects.

**Comment by DanielSchwiperich** [comment]
Please keep also in mind there's the open PR for Multilogin which introduces also some Symfony related changes to the authentication system. There may be some overlappings.
#481

File: src/Resources/contao/classes/BackendUser.php [hunk 1]
```diff
@@ -154,39 +158,6 @@ public function __get($strKey)
 		return parent::__get($strKey);
 	}
 
-
-	/**
-	 * Redirect to the login screen if authentication fails
-	 *
-	 * @return boolean True if the user could be authenticated
-	 */
-	public function authenticate()
```
Comment (discordier) [inline src/Resources/contao/classes/BackendUser.php]: Sadly you must keep this method as it is part of the API and must continue to work as it does currently.

Comment (Toflar) [inline src/Resources/contao/classes/BackendUser.php]: Imho it does because the parent class handles it.

File: src/Resources/contao/library/Contao/User.php [hunk 2]
```diff
@@ -256,58 +260,24 @@ public function getData()
...
-		}
+		/** @var TokenStorageInterface $tokenStorage */
+		$tokenStorage = $container->get('security.token_storage');
 
-		$time = time();
-		$container = \System::getContainer();
-		$session = $container->get('session');
-
-		// Validate the session
-		if ($objSession->sessionID != $session->getId() || (!$container->getParameter('contao.security.disable_ip_check') && $objSession->ip != $this->strIp) || $objSession->hash != $this->strHash || ($objSession->tstamp + \Config::get('sessionTimeout')) < $time)
```
Comment (discordier) [inline src/Resources/contao/library/Contao/User.php]: Restricting the session to an IP must be kept somehow.

File: src/Resources/contao/controllers/BackendIndex.php [hunk 3]
```diff
@@ -32,35 +45,12 @@ class BackendIndex extends \Backend
...
 
-		// Login
-		if ($this->User->login())
-		{
-			$strUrl = 'contao/main.php';
-
-			// Redirect to the last page visited
-			if (\Input::get('referer', true) != '')
-			{
-				$strUrl = base64_decode(\Input::get('referer', true));
```
Comment (discordier) [inline src/Resources/contao/controllers/BackendIndex.php]: Can we pass this information to the login route and redirect back post authentication?

File: src/Resources/contao/library/Contao/User.php (hunk 2 above)
Comment (Toflar) [inline src/Resources/contao/library/Contao/User.php]: No, we said Symfony doesn't so we won't either.

Comment (discordier) [inline src/Resources/contao/library/Contao/User.php]: Which is an BC break then and the whole stuff has to get postponed to 5.0 then... I do not like this.

File: src/Resources/contao/classes/BackendUser.php (hunk 1 above)
Comment (discordier) [inline src/Resources/contao/classes/BackendUser.php]: The redirect is missing in the parent class... in fact, the redirect is the only thing which made `BackendUser::authenticate()` almost unusable for the outside world, however this is an BC break if this method now will behave different.
We do not have to use it in the core anymore but it has to remain for existing extensions which call it and expect the redirect and script termination to happen.

File: src/Resources/contao/library/Contao/User.php (hunk 2 above)
Comment (Toflar) [inline src/Resources/contao/library/Contao/User.php]: This is not a BC break to me.

File: src/Resources/contao/controllers/BackendIndex.php (hunk 3 above)
Comment (bytehead) [inline src/Resources/contao/controllers/BackendIndex.php]: I'll give it a shot :)

File: src/Resources/contao/classes/BackendUser.php (hunk 1 above)
Comment (bytehead) [inline src/Resources/contao/classes/BackendUser.php]: The method is back in there - I've got it wrong in the first view. 
If we now use `BackendUser::authenticate()` we are compatible to all other backend entry points (see https://github.com/contao/core-bundle/pull/559/commits/58e6b65c9569fd70532332118dbdb7baac834bcf).

File: src/Resources/contao/controllers/BackendIndex.php (hunk 3 above)
Comment (bytehead) [inline src/Resources/contao/controllers/BackendIndex.php]: This is probably done [automatically](http://symfony.com/doc/current/security/form_login.html#redirecting-after-success).

First tests worked ok. @discordier Is there any edge cases I could have missed?

Comment (discordier) [inline src/Resources/contao/controllers/BackendIndex.php]: The problem with the original symfony approach is, that it is stored in the session and not as URL parameter.
There might arise issues from this when opening a browser with multiple tabs.
The stored value in the session will be the one for the last tab that get's redirected to the login form then and cause all tabs to contain the same backend route after login.
We could implement it as stated here http://symfony.com/doc/current/security/form_login.html#control-the-redirect-url-from-inside-the-form
We would need to make our value in the template variable though as someone might reconfigure the application to use another parameter name (see config.yml in above link).

Comment (aschempp) [inline src/Resources/contao/controllers/BackendIndex.php]: @bytehead can you check my old implementation. Pretty sure I fixed it there, but I can't remember how exactly.

**Comment by leofeyer** [comment]
@bytehead What's the status on this one?

File: src/Resources/contao/controllers/BackendIndex.php (hunk 3 above)
Comment (bytehead) [inline src/Resources/contao/controllers/BackendIndex.php]: @aschempp didn't find anything. Do you have any hint?

**Comment by bytehead** [comment]
@leofeyer still in progress - didn't find any time during holidays...

File: src/Resources/contao/controllers/BackendIndex.php (hunk 3 above)
Comment (aschempp) [inline src/Resources/contao/controllers/BackendIndex.php]: Here you go: http://symfony.com/doc/current/security/form_login.html#control-the-redirect-url-from-inside-the-form

Comment (bytehead) [inline src/Resources/contao/controllers/BackendIndex.php]: I'm actually curious about your old implementation you've mentioned.

Comment (aschempp) [inline src/Resources/contao/controllers/BackendIndex.php]: That was the solution I remembered, but I could not find my implementation either.

**Comment by aschempp** [comment]
I guess this was not supposed to be closed, but @bytehead and me wanted to create a new PR anyway, right?

**Comment by bytehead** [comment]
Closed in favor of #685.
